### PR TITLE
Tudor/reconnect new heads

### PR DESCRIPTION
**Pull request by tudor-malene**
### Why this change is needed

The `newHeads` subscription is used to invalidate the Ten Gateway cache. It needs to be resilient.

### What changes were made as part of this PR

- introduce reconnecting logic in the "NewHeadService"
- introduce timeout checking in the "NewHeadService" - to handle the case when the backend dies silently
- separate the newHead subscribe logic to pass it as a lambda

### PR checks pre-merging

Please indicate below by ticking the checkbox that you have read and performed the required
[PR checks](https://github.com/ten-protocol/ten-internal/blob/main/dev-ops-docs/dev-pr-checks.md)

- [ ] PR checks reviewed and performed

### CODE REVIEW

File: go/common/subscription/new_heads_manager.go [hunk 1]
```diff
@@ -42,38 +57,60 @@ func NewNewHeadsService(inputCh chan *common.BatchHeader, convertToEthHeader boo
...
+			// for each new head, notify all registered subscriptions
+			for id, notifier := range nhs.newHeadNotifiers {
+				if nhs.stopped.Load() {
+					return nil
+				}
+				err := notifier.Notify(id, msg)
+				if err != nil {
+					// on error, remove the notification
+					nhs.logger.Info("failed to notify newHead subscription", log.ErrKey, err, log.SubIDKey, id)
+					nhs.notifiersMutex.Lock()
```
Comment (BedrockSquirrel) [inline go/common/subscription/new_heads_manager.go:96]: This would deadlock I think, you can't take the write-lock while it's under read-lock. Need to copy the notifiers list before iterating or make list of ones to delete after maybe.

File: go/common/subscription/utils.go [hunk 2]
```diff
@@ -11,7 +11,7 @@ import (
 // ForwardFromChannels - reads messages from the input channels, and calls the `onMessage` callback.
 // Exits when the unsubscribed flag is true.
 // Must be called as a go routine!
-func ForwardFromChannels[R any](inputChannels []chan R, unsubscribed *atomic.Bool, onMessage func(R) error) {
+func ForwardFromChannels[R any](inputChannels []chan R, unsubscribed *atomic.Bool, onMessage func(R) error, onBackendDisconnect func(), timeoutInterval time.Duration) {
```
Comment (BedrockSquirrel) [inline go/common/subscription/utils.go:14]: I'd maybe call the function `onBackendDisconnect` -> `onTimeout` to be clearer, I think that's the only time it's called.

**Review by BedrockSquirrel — COMMENTED** [review]
I think this would do the trick but it's quite hard to follow so I'm not 100% there are no edge cases lurking. Would consider the approach we discussed in discord but I guess either way it's going to be a bit hairy lol. I can approve this if you want to stick with it (and fix up the deadlock thing below).

File: go/common/subscription/new_heads_manager.go (hunk 1 above)
Comment (tudor-malene) [reply go/common/subscription/new_heads_manager.go:96]: I haven't changed this bit. Just reformatted. It might have been there before. I'll have a look

File: go/common/subscription/utils.go (hunk 2 above)
Comment (tudor-malene) [reply go/common/subscription/utils.go:14]: It is also called when the backend disconnects via the usual means.

Comment (BedrockSquirrel) [reply go/common/subscription/utils.go:14]: Usual means meaning when `unsubscribed` is set to true? Think you need to change a `return` to a `break` then in the for loop

Comment (tudor-malene) [reply go/common/subscription/utils.go:14]: this bit here:
```	for unclosedInputChannels > 0 {
		chosen, value, ok := reflect.Select(inputCases)
		if !ok {
			// The chosen channel has been closed, so zero out the channel to disable the case
			inputCases[chosen].Chan = reflect.ValueOf(nil)
			unclosedInputChannels--
			continue
		}
```

I'll add a comment to clarify it

File: go/common/subscription/new_heads_manager.go [hunk 3]
```diff
@@ -93,18 +78,14 @@ func (nhs *NewHeadsService) Start() error {
...
 			}
 			return nil
 		},
 		func() {
-			if nhs.connectFunc == nil {
-				nhs.logger.Crit("the inbound new heads channel was closed.")
-			}
-			err := nhs.connectFunc(nhs.inputCh)
+			nhs.logger.Info("Disconnected from new head subscription. Reconnecting...")
+			inputCh, err = nhs.connectFunc()
```
Comment (BedrockSquirrel) [inline go/common/subscription/new_heads_manager.go:88]: This assignment doesn't have the desired effect I don't think. It won't update the channel in the slice that was passed into ForwardFromChannels above.

To that you'd need to reference the slice, something like:
```
inputChans := []chan *common.BatchHeader{inputCh}

// and then down here you can write:			
inputCh, err = nhs.connectFunc()
inputChans[0] = inputCh
```

I wasn't sure how this worked, so did this to convince myself lol: https://go.dev/play/p/YD7x0HaUtoO

Comment (BedrockSquirrel) [reply go/common/subscription/new_heads_manager.go:88]: Actually even that fix might be problematic I think (if I'm reading the magical incantations correctly).

The `inputCases` you build up with reflection at the start of `ForwardFromChannels()` will still be listening on the old channel.
